src/ERC1271.sol

```diff
@@ -1,50 +1,77 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.4;
 
-/// @notice ERC-1271 with guards for same signer being used on multiple accounts
-/// Based on Solady (https://github.com/vectorized/solady/blob/main/src/accounts/ERC1271.sol)
+/// @title ERC-1271 With Cross Account Replay Protection
+///
+/// @notice Abstract ERC-1271 implementation (based on Solady's one) with guards to handle the same
+///         signer being used on multiple accounts, using a nested approach based on EIP-712.
+///
+/// @dev To prevent the same signature from being validated on different accounts owned by the samer signer,
+///      we introduce an anti cross-account-replay layer: the original signed hash (OSH) is wrapped in a
+///      `CoinbaseSmartWalletMessage` (CSWM) struct, which is itself hashed using EIP-712.
+///
+///      When hashing the CSWM, the `domainSeparator` used has its `verifyingContract` field set to the address
+///      of the targeted account. This mechanism is coupling the signature of the CSWM hash with the account,
+///      effectively preventing it from being replayed on another account owned by the same signer.
+///
+///      See `replaySafeHash()` for the implementation details.
+///
 /// @author Coinbase (https://github.com/coinbase/smart-wallet)
+/// @author Solady (https://github.com/vectorized/solady/blob/main/src/accounts/ERC1271.sol)
 abstract contract ERC1271 {
-    /// @dev We use `bytes32 hash` rather than `bytes message`
-    /// In the EIP-712 context, `bytes message` would be useful for showing users a full message
-    /// they are signing in some wallet preview. But in this case, to prevent replay
-    /// across accounts, we are always dealing with nested messages, and so the
-    /// input should be a EIP-191 or EIP-712 output hash.
-    /// E.g. The input hash would be result of
+    /// @dev Precomputed `typeHash` used to produce the CSWM hash using EIP-712.
     ///
-    ///  keccak256("\x19\x01" || someDomainSeparator || hashStruct(someStruct))
+    ///      The OSH must either be:
```

Comment:
It's not signed, it's just "original hash"

Reply:
Not sure the abbreviations are making things more clear.

Reply:
Done in #2e6f687. I kept CSWM abbreviations at it's a lot smaller to write but can remove it wanted.